pkg/stats/realtime.go

```diff
@@ -82,6 +87,74 @@ func loopJSON(client api.RealtimeStatsInterface, service string, out io.Writer)
 }
 
 func loopText(client api.RealtimeStatsInterface, service string, out io.Writer) error {
+	responses := make(chan realtimeResponse)
+	errors := make(chan error)
+	go fetchData(client, service, responses, errors)
+
+	for {
+		select {
+		case err := <-errors:
+			text.Error(out, "fetching stats: %w", err)
+			continue
+		case envelope := <-responses:
+			for _, block := range envelope.Data {
+				agg := block.Aggregated
+
+				// FIXME: These are heavy-handed compatibility
+				// fixes for stats vs realtime, so we can use
+				// fmtBlock for both.
+				agg["start_time"] = block.Recorded
+				delete(agg, "miss_histogram")
+
+				if err := fmtBlock(out, service, agg); err != nil {
+					text.Error(out, "formatting stats: %w", err)
+					continue
+				}
+			}
+		}
+	}
+}
+
+func loopChart(client api.RealtimeStatsInterface, service string, out io.Writer) error {
+	responses := make(chan realtimeResponse)
+	errors := make(chan error)
+
+	view, err := NewView(service)
+	if err != nil {
+		return err
+	}
+	defer view.Close()
+	view.Resize()
+	view.SetLayout()
+
+	go fetchData(client, service, responses, errors)
+
+	for {
+		select {
+		case err := <-errors:
```

Comment:
Ok, here it makes more sense to use a channel because you do want to `select` across your remote calls and `view` events (because the termui API is all channel based).

So I might do a combination of what was above:
* make one channel of `struct { realtimeResponse, error }` (to make future loop changes easier)
* have the various loopFoos create a `context.Context` using WithCancel so you get a CancelFunc
* make fetchData cancel-able by adding a `ctx context.Context` as its first argument (and check for cancelation in its loop)
* call the CancelFunc at all of the loopFoo exit points (defer is fine here too) so that fetchData won't survive any longer than loopFoo